domaci_maze.py

```diff
@@ -90,495 +77,523 @@ def has_value(self) -> bool:
-            raise Exception("Invalid row position.")
-        if col < 0 or col >= self.cols_no:
-            raise Exception("Invalid column position.")
-        if not self.board[row, col].is_steppable():
-            raise Exception("Invalid position: unsteppable cell.")
-        return row, col
-
-    def move_from(self, row: int, col: int, action: str) -> tuple[int, int]:
-        if action in self.graph[self.board[row, col]]:
-            return self.graph[self.board[row, col]][action]
-        else:
-            raise Exception(f"Invalid action: {action} for cell ({row}, {col}).")
-
-    def __call__(
-        self, state: tuple[int, int], action: str
-    ) -> tuple[tuple[int, int], float, bool]:
-        row, col = state
-        new_row, new_col = self.move_from(row, col, action)
-        new_cell = self.board[new_row, new_col]
-        reward = new_cell.get_reward()
-        is_terminal = new_cell.is_terminal()
-        return (new_row, new_col), reward, is_terminal
-
-    def get_states(self):
-        states = []
-        for r in range(self.board.rows_no):
-            for c in range(self.board.cols_no):
-                if self.board[r, c].is_steppable():
-                    states.append((r, c))
-        return states
-
-    def is_terminal(self, s):
-        return self.board[s[0], s[1]].is_terminal()
-
-    def get_actions(self, cell):
-        if cell in self.graph:
-            return list(self.graph[cell].keys())
-        else:
+    def set_probabilities(self, node, g):
+
+        # terminal and wall node do not have any probabilities
+        if isinstance(node, WallNode) or isinstance(node, TerminalNode):
             return []
 
+        # list is easier to iterate, only because of that is node_list made
+        nodes_list = [node for node in g]
+        nodes_list_help = copy(nodes_list)  # only for iteration
+
+        probabilities = []
+
+        total_cells = self.graph_width * self.graph_height
+        zero_cells = total_cells // 2  # number of cells with zero prob of stepping
+
+        # probability is set to zero for wall modes
+        for n in nodes_list_help:
+            if isinstance(n, WallNode):
+                probabilities.append([0, n, 0])
+                zero_cells -= 1
+                nodes_list.remove(n)
+
+        # for regular nodes, half of probabilities should be zero, including wall nodes
+        if isinstance(node, RegularNode):
+            while zero_cells != 0:
+                random_node = rdm.choice(nodes_list)
+                nodes_list.remove(random_node)
+                probabilities.append([0, random_node, 0])
+                zero_cells -= 1
+
+        # number of nodes that can be reached with one action
+        # action_len = math.ceil(len(nodes_list) / len(ACTIONS))
+        if len(nodes_list) / len(ACTIONS) > 1.5:
+            action_len = math.ceil(len(nodes_list) / len(ACTIONS))
+        else:
+            action_len = math.floor(len(nodes_list) / len(ACTIONS))

```

Comment:
Umesto ovoga si mogao samo reci round i to ti zaokruzi na najblizu vrednost, ovo mi je bas komplikacija